lib/acl.c

```diff
@@ -0,0 +1,224 @@
+		.input_index = PROTO_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, proto),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST1_FIELD_IPV6,
+		.input_index = DST1_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[0]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST2_FIELD_IPV6,
+		.input_index = DST2_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[4]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST3_FIELD_IPV6,
+		.input_index = DST3_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[8]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST4_FIELD_IPV6,
+		.input_index = DST4_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[12]),
+	},
+	{
+		/* Enforce grouping into four bytes. */
+		.type = RTE_ACL_FIELD_TYPE_BITMASK,
+		.size = sizeof(uint32_t),
+		.field_index = TYPE_FIELD_ICMPV6,
+		.input_index = TYPE_FIELD_ICMPV6,
+		.offset = sizeof(struct ipv6_hdr) +
+			offsetof(struct icmpv6_hdr, type),
+	},
+};
+
+int
+register_ipv6_acl(struct ipv6_acl_rule *ipv6_rules, unsigned int num_rules,
+	acl_cb_func cb_f, struct gatekeeper_if *iface)
+{
+	unsigned int numa_nodes = get_net_conf()->numa_nodes;
+	unsigned int i;
+
+	if (iface->acl_count == GATEKEEPER_IPV6_ACL_MAX) {
+		RTE_LOG(ERR, GATEKEEPER, "acl: cannot install more rules on the %s iface\n",
+			iface->name);
+		return -1;
+	}
+
+	/* Assign a new ID for this rule type. */
+	for (i = 0; i < num_rules; i++)
+		ipv6_rules[i].data.userdata = iface->acl_count;
+
+	for (i = 0; i < numa_nodes; i++) {
+		int ret = rte_acl_add_rules(iface->ipv6_acls[i],
+			(struct rte_acl_rule *)ipv6_rules, num_rules);
+		if (ret < 0) {
+			RTE_LOG(ERR, ACL, "Failed to add ACL rules on the %s interface on socket %d\n",
+				iface->name, i);
+			return ret;
+		}
+	}
+
+	iface->acl_funcs[iface->acl_count] = cb_f;
+	iface->acl_count++;
+
+	return 0;
+}
+
+int
+process_ipv6_acl(struct gatekeeper_if *iface, unsigned int lcore_id,
+	struct acl_search *acl)
+{
+	struct rte_mbuf *pkts
+		[GATEKEEPER_IPV6_ACL_MAX][GATEKEEPER_MAX_PKT_BURST];
+	int num_pkts[GATEKEEPER_IPV6_ACL_MAX] = { 0 };
+	unsigned int socket_id = rte_lcore_to_socket_id(lcore_id);
+	int i;
+
+	int ret = rte_acl_classify(iface->ipv6_acls[socket_id],
+		acl->data, acl->res, acl->num, 1);
+	if (ret < 0) {
+		RTE_LOG(ERR, ACL,
+			"invalid arguments given to rte_acl_classify()\n");
+		drop_ipv6_acl_pkts(acl->mbufs, acl->num, iface);
+		return ret;
+	}
+
+	/* Split packets into separate buffers -- one for each type. */
+	for (i = 0; i < acl->num; i++) {
+		int type = acl->res[i];
+		pkts[type][num_pkts[type]++] = acl->mbufs[i];
+	}
+
+	/* Transmit separate buffers to registered ACL functions. */
+	for (i = 0; i < GATEKEEPER_IPV6_ACL_MAX; i++) {
+		ret = ((acl_cb_func)iface->acl_funcs[i])(pkts[i],
+			num_pkts[i], iface);
```

Comment:
`submit_nd()` does a lot of nothing when `num_pkts[i] == 0`, and this behavior is kind of expected to happen in other callback functions, so adding a test `num_pkts[i] > 0` before calling the callback function is a good idea.